python/mcap-ros2-support/mcap_ros2/dynamic.py

```diff
@@ -196,28 +290,246 @@ def _read_complex_type(
+                    if ftype.is_fixed_size_array() and ftype.array_size is not None
+                    else reader.sequence_length()
+                )
+                value = array_parser_fn(reader, array_length)
                 setattr(msg, field.name, value)
             else:
-                parser_fn = FIELD_PARSERS[field.type.type]
+                parser_fn = FIELD_PARSERS[ftype.type]
                 if parser_fn is None:
                     raise NotImplementedError(
-                        f"Parsing for type {field.type.type} is not implemented"
+                        f"Parsing for type {ftype.type} is not implemented"
                     )
                 value = parser_fn(reader)
                 setattr(msg, field.name, value)
 
     return msg
 
 
+def _write_complex_type(
+    msg_name: str,
+    fields: List[Field],
+    msgdefs: Dict[str, MessageSpecification],
+    ros2_msg: Any,
+    writer: CdrWriter,
+) -> None:
+    for field in fields:
+        ftype = field.type
+        if not ftype.is_primitive_type():
+            # Complex type
+            nested_definition = msgdefs[f"{ftype.pkg_name}/{ftype.type}"]
+            if nested_definition is None:
+                raise ValueError(
+                    f'Message definition not found for field "{field.name}" with '
+                    'type "{ftype.type}"'
+                )
+
+            if ftype.is_array:
+                array: Union[List[Any], Any] = _get_property(ros2_msg, field.name)
+                if array is None:
+                    array = []
+                if not isinstance(array, list):
+                    raise ValueError(
+                        f'Field "{field.name}" is not an array but has array type '
+                        f'"{ftype.type}[]"'
+                    )
+
+                if ftype.is_fixed_size_array() and ftype.array_size is not None:
+                    # Fixed length array, ensure the input array is the correct length
+                    while len(array) < ftype.array_size:
+                        array.append({})
+                    if len(array) > ftype.array_size:
+                        array = array[: ftype.array_size]
+
+                    for item in array:
+                        _write_complex_type(
+                            nested_definition.msg_name,
+                            nested_definition.fields,
+                            msgdefs,
+                            item,
+                            writer,
+                        )
+                else:
+                    # Limit the array to the upper bound length, if present
+                    if (
+                        ftype.is_upper_bound
+                        and ftype.array_size is not None
+                        and len(array) > ftype.array_size
+                    ):
+                        array = array[: ftype.array_size]
+
+                    # Dynamic length array, write a uint32 prefix
```

Comment:
Is the encoding of variable length arrays specified as part of CDR or just a convention that ROS 2 uses on top of it? If the former, maybe this could move into the writer class